### DISCUSSION TITLE
seed not work on with pnpx

### DISCUSSION THREAD

**Discussion by howarddo2208** [prisma/prisma#19111] [Q&A] [answered]
### Bug description

I'm building a next.js app using prisma and pnpm as my package manager. Somehow the seed script only execute successfully with npx: `npx prisma db seed` but not pnpx when I run `pnpx prisma db seed`
Here is the script run log:
```
.../Library/pnpm/store/v3/tmp/dlx-30360  | Progress: resolved 1, reused.../Library/pnpm/store/v3/tmp/dlx-30360  |   +2 +
.../Library/pnpm/store/v3/tmp/dlx-30360  | Progress: resolved 1, reused.../Library/pnpm/store/v3/tmp/dlx-30360  | Progress: resolved 2, reused.../Library/pnpm/store/v3/tmp/dlx-30360  | Progress: resolved 2, reused 2, downloaded 0, added 2, done
Environment variables loaded from .env
Running seed command `ts-node --compiler-options {"module":"CommonJS"} prisma/seed.ts` ...

An error occurred while running the seed command:
Error: Command failed with ENOENT: ts-node --compiler-options {"module":"CommonJS"} prisma/seed.ts
spawn ts-node ENOENT
 ERROR  Command failed with exit code 1: prisma db seed

pnpm: Command failed with exit code 1: prisma db seed
    at makeError (/Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:24230:17)
    at handlePromise (/Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:24801:33)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.handler [as dlx] (/Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:206754:7)
    at async /Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:216086:21
    at async main (/Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:216053:34)
    at async runPnpm (/Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:216308:5)
    at async /Users/tanhopdo/.cache/node/corepack/pnpm/8.3.1/dist/pnpm.cjs:216300:7
```

### How to reproduce

1. run `pnpx prisma db seed`
2. See error
```json
// package.json
{
  "name": "web-shop",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@prisma/client": "^4.13.0",
    "@types/node": "20.0.0",
    "@types/react": "18.2.5",
    "@types/react-dom": "18.2.3",
    "@vercel/postgres": "^0.1.3",
    "autoprefixer": "10.4.14",
    "bcryptjs": "^2.4.3",
    "eslint": "8.39.0",
    "eslint-config-next": "13.4.0",
    "next": "13.4.0",
    "next-auth": "^4.22.1",
    "postcss": "8.4.23",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "tailwindcss": "3.3.2",
    "ts-node": "^10.9.1",
    "typescript": "5.0.4"
  },
  "devDependencies": {
    "@types/bcryptjs": "^2.4.2",
    "prisma": "^4.13.0"
  },
  "prisma": {
    "seed": "ts-node --compiler-options {\"module\":\"CommonJS\"} prisma/seed.ts"
  }
}
```

### Expected behavior

be able to seed whether with `npx` or `pnpx`

### Prisma information

```prisma
// schema.prisma

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider          = "postgresql"
  url               = env("POSTGRES_PRISMA_URL") // uses connection pooling
  directUrl         = env("POSTGRES_URL_NON_POOLING") // uses a direct connection
}

model User {
  id       String @id @default(uuid())
  name     String
  email    String @unique
  password String
}
```

```ts
// seed.ts
import { PrismaClient } from "@prisma/client";
import { hash } from "bcryptjs";

const prisma = new PrismaClient();

async function main() {
  const password = await hash("password123", 12);
  const user = await prisma.user.upsert({
    where: { email: "admin@admin.com" },
    update: {},
    create: {
      email: "admin@admin.com",
      name: "Admin",
      password,
    },
  });
  console.log({ user });
}
main()
  .then(() => prisma.$disconnect())
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });
```

### Environment & setup

- OS: macOS
- Database: Postgres (Vercel Postgres)
- Node.js version: 18.16.0
- pnpm version: 8.3.1

### Prisma Version

```
4.13.0
```

**Comment by ludralph** [comment, marked as answer]
Hi @howarddo2208 👋 

`pnpx` is deprecated. See this Stackoverflow [post](https://stackoverflow.com/questions/70248239/how-to-install-pnpx). It is recommended to use  [pnpm dlx](https://pnpm.io/cli/dlx) and [pnpm exec](https://pnpm.io/cli/exec) instead of `pnpx`.

**Reply by howarddo2208** [reply]
Oh, thank u, I didn't know that

**Reply by serdarsen** [reply]
> Hi @howarddo2208 wave
> 
> `pnpx` is deprecated. See this Stackoverflow [post](https://stackoverflow.com/questions/70248239/how-to-install-pnpx). It is recommended to use [pnpm dlx](https://pnpm.io/cli/dlx) and [pnpm exec](https://pnpm.io/cli/exec) instead of `pnpx`.

Thank you for bringing to our attention that `pnpx` is deprecated. It is recommended to use `pnpm dlx` instead. 

I tried `pnpm dlx` on [Fullstack Authentication Example with Next.js and NextAuth.js](https://github.com/prisma/prisma-examples/tree/latest/typescript/rest-nextjs-api-routes-auth#fullstack-authentication-example-with-nextjs-and-nextauthjs). 

`pnpx prisma db seed` and `pnpm dlx prisma db seed` didn't work. `npx prisma db seed` worked. 

I'm on Ubuntu 22.04.2 LTS, npm 9.6.4, pnpm 8.5.1, npx 9.6.4, node v20.1.0. I only chanced datasource db to postgresql. I cloned the repo successfully using `pnpx try-prisma@latest --template typescript/rest-nextjs-api-routes-auth` (I didn't know that `pnpx` is deprecated. It didn't automatically seed either).

Update:
I cloned the repo from scratch successfully using `pnpm dlx  try-prisma@latest --template typescript/rest-nextjs-api-routes-auth`. I migrated the tables successfully using `pnpm dlx prisma migrate dev --name init`. It didn't automatically seed.  `pnpm dlx prisma db seed` didn't work again. You can see the following errors.

```bash
srdr@asus:~/dev/rest-nextjs-api-routes-auth$ pnpm dlx prisma db seed
.../share/pnpm/store/v3/tmp/dlx-15547    |   +2 +
.../share/pnpm/store/v3/tmp/dlx-15547    | Progress: resolved 2, reused 2, downloaded 0, added 2, done
Environment variables loaded from .env
Running seed command `ts-node prisma/seed.ts` ...

An error occurred while running the seed command:
Error: Command failed with ENOENT: ts-node prisma/seed.ts
spawn ts-node ENOENT
 ERROR  Command failed with exit code 1: prisma db seed

pnpm: Command failed with exit code 1: prisma db seed
    at makeError (/home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:24242:17)
    at handlePromise (/home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:24813:33)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Object.handler [as dlx] (/home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:223404:7)
    at async /home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:232806:21
    at async main (/home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:232773:34)
    at async runPnpm (/home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:233028:5)
    at async /home/srdr/.cache/node/corepack/pnpm/8.5.1/dist/pnpm.cjs:233020:7

```

**Reply by janpio** [reply]
Hm, `ts-node` is a dev dependency so should be present - but the error message `ENOENT: ts-node prisma/seed.ts` says it is not.
Does `pnpm dlx ts-node prisma/seed.ts` manually work?

**Reply by serdarsen** [reply]
> Hm, `ts-node` is a dev dependency so should be present - but the error message `ENOENT: ts-node prisma/seed.ts` says it is not. Does `pnpm dlx ts-node prisma/seed.ts` manually work?

Yes, thank you @janpio! The command `pnpm dlx ts-node prisma/seed.ts` executed successfully. Here is the output for reference:
```bash
srdr@asus:~/dev/rest-nextjs-api-routes-auth$ pnpm dlx ts-node prisma/seed.ts
.../share/pnpm/store/v3/tmp/dlx-14975    |  +19 ++
.../share/pnpm/store/v3/tmp/dlx-14975    | Progress: resolved 19, reused 18, downloaded 1, added 19, done
Start seeding ...
Created user with id: 4
Created user with id: 5
Created user with id: 6
Seeding finished.
```

**Reply by janpio** [reply]
Weird. With `npx`, running `npx prisma db seed` takes the `ts-node prisma/seed.ts` command and runs it internally as `npx ts-node prisma/seed.ts`. That does not seem to be the case for `pnpm dlx`.

Can you open a new bug issue about this please? Thanks.

**Reply by serdarsen** [reply]
Thank you, @janpio, for your explanation and assistance. It's interesting to learn that the behavior of `npx` and `pnpm dlx` commands differs in this regard. I appreciate you pointing it out.

I will take your suggestion and open a new bug issue to report this discrepancy. I'll make sure to include the relevant details and context from our discussion to provide a clear explanation of the problem.

[rest-nextjs-api-routes-auth pnpm dlx prisma db seed issue](https://github.com/prisma/prisma-examples/issues/4749)
